bootstrap/internal/ignition/ignition.go

```diff
@@ -0,0 +1,182 @@
+	}
+
+	workerSystemdServices = []string{
+		"systemctl enable rke2-agent.service",
+		"systemctl start rke2-agent.service",
+	}
+)
+
+// JoinWorkerInput defines the context to generate a node user data.
+type JoinWorkerInput struct {
+	*cloudinit.BaseUserData
+
+	AdditionalIgnition *bootstrapv1.AdditionalUserData
+}
+
+// ControlPlaneJoinInput defines context to generate controlplane instance user data for control plane node join.
+type ControlPlaneJoinInput struct {
+	*cloudinit.ControlPlaneInput
+
+	AdditionalIgnition *bootstrapv1.AdditionalUserData
+}
+
+// ControlPlaneInitInput defines the context to generate a controlplane instance user data.
+type ControlPlaneInitInput struct {
+	*cloudinit.ControlPlaneInput
+
+	AdditionalIgnition *bootstrapv1.AdditionalUserData
+}
+
+// NewJoinWorker returns Ignition configuration for new worker node joining the cluster.
+func NewJoinWorker(input *JoinWorkerInput) ([]byte, error) {
+	if input == nil {
+		return nil, fmt.Errorf("input can't be nil")
+	}
+
+	if input.BaseUserData == nil {
+		return nil, fmt.Errorf("node input can't be nil")
+	}
+
+	deployRKE2Command, err := getWorkerRKE2Commands(input.BaseUserData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get rke2 command: %w", err)
+	}
+
+	input.DeployRKE2Commands = deployRKE2Command
+	input.WriteFiles = append(input.WriteFiles, input.ConfigFile)
+
+	return render(input.BaseUserData, input.AdditionalIgnition)
+}
+
+// NewJoinControlPlane returns Ignition configuration for new controlplane node joining the cluster.
+func NewJoinControlPlane(input *ControlPlaneJoinInput) ([]byte, error) {
+	if input == nil {
+		return nil, fmt.Errorf("input can't be nil")
+	}
+
+	if input.ControlPlaneInput == nil {
+		return nil, fmt.Errorf("controlplane join input can't be nil")
+	}
+
+	deployRKE2Command, err := getControlPlaneRKE2Commands(&input.BaseUserData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get rke2 command: %w", err)
+	}
+
+	input.DeployRKE2Commands = deployRKE2Command
+	input.WriteFiles = append(input.WriteFiles, input.Certificates.AsFiles()...)
+	input.WriteFiles = append(input.WriteFiles, input.ConfigFile)
+
+	return render(&input.BaseUserData, input.AdditionalIgnition)
+}
+
+// NewInitControlPlane returns Ignition configuration for bootstrapping new cluster.
+func NewInitControlPlane(input *ControlPlaneInitInput) ([]byte, error) {
+	if input == nil {
+		return nil, fmt.Errorf("input can't be nil")
+	}
+
+	if input.ControlPlaneInput == nil {
+		return nil, fmt.Errorf("controlplane input can't be nil")
+	}
+
+	deployRKE2Command, err := getControlPlaneRKE2Commands(&input.BaseUserData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get rke2 command: %w", err)
+	}
+
+	input.DeployRKE2Commands = deployRKE2Command
+	input.WriteFiles = append(input.WriteFiles, input.Certificates.AsFiles()...)
+	input.WriteFiles = append(input.WriteFiles, input.ConfigFile)
+
+	return render(&input.BaseUserData, input.AdditionalIgnition)
+}
```

Comment:
Theres quite a bit of duplicate code in these functions. Makes me think if we could have 1 function?

Reply:
That makes sense, I will extract duplicate code in a separate function and leave NewJoinControlPlane and NewInitControlPlane in case we will want to customize control plane configurations